connection.go

```diff
@@ -162,6 +177,13 @@ type Connection struct {
 	serverProtocolInfo ProtocolInfo
 	// watchMap is a map of key -> chan watchState.
 	watchMap sync.Map
+
+	// shutdownInProgress defined whether shutdown now in progress.
+	shutdownInProgress uint32
+	// shutdownWatcher is the "box.shutdown" event watcher.
+	shutdownWatcher Watcher
```

Comment:
https://github.com/tarantool/go-tarantool/blob/265f96c9ff5f39e436e7645845fea4589762174d/connection.go#L1483-L1485

Reply:
Direct unregister seems potentially dangerous: server waits until all subscribed connections finish on graceful shutdown, but if we unsubscribe our connection, it is expected to (I'm not checked yet) fall out of graceful shutdown procedure.

I have talked with @locker about net.box approach and he said that unregister not needed here.

Reply:
"Socket close is a session destruction, and it leads to all events unwatch."

Reply:
> "Socket close is a session destruction, and it leads to events unwatch."

The watcher goroutine is still run if we don't call Unregister(). The note is about the Go API.

Reply:
Can we kill the goroutine without sending UNWATCH request?

Reply:
I don't see a way to Unregister from an event callback:

- `Unregister` waits for `finished`
-
  > 	// finished is closed when the watcher is unregistered and the watcher
  > 	// goroutine is finished.

- watcher goroutine doesn't finish until callback returnes something

and we've got a deadlock.

Reply:
Btw, how does this one work with reconnect? Does missing Unregister breaks inner counter of active watchers? If I add Unregister in some way, will I be required to create new watcher each time on reconnect (since this one is no longer active)?

Reply:
I did not say that you need to call Unregister() in the callback:

https://github.com/tarantool/go-tarantool/blob/265f96c9ff5f39e436e7645845fea4589762174d/connection.go#L1487-L1488